docs(testing): add componentOnReady guidance and a Vue testing page (#4611)
ionic-team/ionic-docs: docs/react/testing/unit-testing/best-practices.md

```

For more information on `user-event`, see the [user-event documentation](https://testing-library.com/docs/user-event/intro/).

> ## Waiting for Components
> 
> When you need to wait for an Ionic component to render before asserting against its DOM, use the `componentOnReady` helper exported from `@ionic/core`. Do not call `el.componentOnReady()` directly. `@ionic/react` uses Stencil's custom elements build, where that method does not exist on the element. The helper waits one animation frame instead, giving the component's inner contents a chance to render.
> 
> ```tsx
> import { test, expect } from 'vitest';
> import { render } from '@testing-library/react';
> import { componentOnReady } from '@ionic/core';
> 
> import App from './App';
> 
> test('renders the submit button', async () => {
>   const { container } = render(<App />);
> 
>   const button = container.querySelector('ion-button');
> 
>   await new Promise<void>((resolve) => componentOnReady(button!, () => resolve()));
> 
>   expect(button?.textContent).toContain('Submit');
> });
> ```

Reviewer: Wrapped in IonApp to match the first best practice on this page, moved the non-null assertion to the query so the last line reads cleanly, and dropped the vitest import since the other snippets here stay runner agnostic.

```suggestion
```tsx
import { IonApp } from '@ionic/react';
import { render } from '@testing-library/react';
import { componentOnReady } from '@ionic/core';

import Example from './Example';

test('renders the submit button', async () => {
  const { container } = render(
    <IonApp>
      <Example />
    </IonApp>
  );

  const button = container.querySelector('ion-button')!;

  await new Promise<void>((resolve) => componentOnReady(button, () => resolve()));

  expect(button.textContent).toContain('Submit');
});
```
```

```

For more information on `user-event`, see the [user-event documentation](https://testing-library.com/docs/user-event/intro/).

## Waiting for Components

When you need to wait for an Ionic component to render before asserting against its DOM, use the `componentOnReady` helper exported from `@ionic/core`. Do not call `el.componentOnReady()` directly. `@ionic/react` uses Stencil's custom elements build, where that method does not exist on the element. The helper waits one animation frame instead, giving the component's inner contents a chance to render.

```tsx
import { IonApp } from '@ionic/react';
import { render } from '@testing-library/react';
import { componentOnReady } from '@ionic/core';

import Example from './Example';

test('renders the submit button', async () => {
  const { container } = render(
    <IonApp>
      <Example />
    </IonApp>
  );

>   const button = container.querySelector('ion-button');

Reviewer: `querySelector` returns `Element | null`, so `button.textContent` on the last line fails under `strictNullChecks`.

```suggestion
  const button = container.querySelector('ion-button')!;
```

```

For more information on `user-event`, see the [user-event documentation](https://testing-library.com/docs/user-event/intro/).

> ## Import `componentOnReady` for standalone projects

Reviewer: ```suggestion
## Waiting for Components
```

> ## Import `componentOnReady` for standalone projects

Reviewer: We should also include an example like Angular. The example should match the one that I provided for Angular.

Author: OK, I tested and added one. Let me know if it needs to be changed.

> When testing Ionic components, use the exported `componentOnReady` helper from `@ionic/core` instead of calling `el.componentOnReady()` directly. The helper works with both lazy-loaded and custom-element builds, making it more likely the component has finished rendering before making assertions against its rendered DOM or running accessibility tests.

Reviewer: `@ionic/react` only ships the custom elements build, so `el.componentOnReady()` is never present and the standalone/both-builds framing does not apply here; this states the one behavior that's true for every React project.

```suggestion
When you need to wait for an Ionic component to render before asserting against its DOM, use the `componentOnReady` helper exported from `@ionic/core`. Do not call `el.componentOnReady()` directly. `@ionic/react` uses Stencil's custom elements build, where that method does not exist on the element. The helper waits one animation frame instead, giving the component's inner contents a chance to render.
```
